synedrion/src/paillier/rsa.rs

```diff
@@ -0,0 +1,282 @@
+            P::MODULUS_BITS,
+        )
+        .expect(concat![
+            "the totient is smaller than the modulus, ",
+            "and thefore can be bounded by 2^MODULUS_BITS"
+        ])
+    }
+}
+
+/// The minimized structure containing the public RSA modulus.
+///
+/// Suitable for serialization or transmission.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub(crate) struct PublicModulusWire<P: PaillierParams>(P::Uint);
+
+impl<P: PaillierParams> PublicModulusWire<P> {
+    fn new(primes: &SecretPrimesWire<P>) -> Self {
+        Self(primes.p.expose_secret().mul_wide(primes.q.expose_secret()))
+    }
+
+    pub fn into_precomputed(self) -> PublicModulus<P> {
+        PublicModulus::new(self)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct PublicModulus<P: PaillierParams> {
+    /// The base RSA modulus $N$.
+    modulus: PublicModulusWire<P>,
+    /// Montgomery representation parameters for modulo $N$.
+    monty_params_mod_n: <P::UintMod as Monty>::Params,
+}
+
+impl<P: PaillierParams> PartialEq for PublicModulus<P> {
+    fn eq(&self, other: &PublicModulus<P>) -> bool {
+        // Only compare the moduli themselves, not the precomputed constants.
+        self.modulus.eq(&other.modulus)
+    }
+}
+
+impl<P: PaillierParams> Eq for PublicModulus<P> {}
+
+impl<P: PaillierParams> PublicModulus<P> {
+    pub fn new(modulus: PublicModulusWire<P>) -> Self {
+        let odd_modulus = Odd::new(modulus.0).expect("the RSA modulus is odd");
+        let monty_params_mod_n = P::UintMod::new_params_vartime(odd_modulus);
+        Self {
+            modulus,
+            monty_params_mod_n,
+        }
+    }
+
+    pub fn to_wire(&self) -> PublicModulusWire<P> {
+        self.modulus.clone()
+    }
+
+    pub fn modulus(&self) -> &P::Uint {
+        &self.modulus.0
+    }
+
+    pub fn modulus_nonzero(&self) -> NonZero<P::Uint> {
+        NonZero::new(self.modulus.0).expect("the modulus is non-zero")
+    }
+
+    pub fn modulus_bounded(&self) -> Bounded<P::Uint> {
+        Bounded::new(self.modulus.0, P::MODULUS_BITS).expect("the modulus can be bounded by 2^MODULUS_BITS")
+    }
+
+    pub fn monty_params_mod_n(&self) -> &<P::UintMod as Monty>::Params {
+        &self.monty_params_mod_n
+    }
+
+    /// Returns a uniformly chosen number in range $[0, N)$ such that it is invertible modulo $N$, in Montgomery form.
+    pub fn random_invertible_group_elem(&self, rng: &mut impl CryptoRngCore) -> P::UintMod {
+        let modulus = self.modulus_nonzero();
+        loop {
+            let r = P::Uint::random_mod(rng, &modulus);
+            let r_m = P::UintMod::new(r, self.monty_params_mod_n.clone());
+            if r_m.invert().is_some().into() {
+                return r_m;
+            }
+        }
+    }
+
+    /// Returns a uniformly chosen quadratic residue modulo $N$, in Montgomery form.
+    pub fn random_square_group_elem(&self, rng: &mut impl CryptoRngCore) -> P::UintMod {
```

Comment:
```suggestion
    pub fn random_quadratic_residue(&self, rng: &mut impl CryptoRngCore) -> P::UintMod {
```
Maybe a better name?